optaplanner-website-root/content/blog/2022/09/01/Bavet-a-faster-score-engine-for-OptaPlanner.adoc

```diff
@@ -0,0 +1,288 @@
+
+Twice as fast score calculation. Zero API changes.
+
+== Faster
+
+For 20 diverse use cases, we compared Bavet and Drools for OptaPlanner score calculation.
+We ran JMH benchmarks
+on OpenJDK 17 (`ParallelGC`, `Xmx1G`)
+on a stable benchmark machine (`Intel® Xeon® Silver (12 cores total / 24 threads)` and `128 GiB` RAM memory)
+without any other computational demanding processes running.
+
+On average, *Bavet is twice as fast as Drools* for score calculation.
+In the https://www.optaplanner.org/learn/useCases/vehicleRoutingProblem.html[Vehicle Routing Problem],
+Bavet is even three times as fast as Drools:
+
+// TODO Update table, put Drools first, percentage -= 100, switch rows/columns?
+
+// TODO Graph!
+
+.Score calculation speed on different use cases
+|===
+|Use case |Bavet |Drools |Speed up +100%
+
+|cheaptime >|18,270 >|7,949 >|230%
+|cloudbalancing >|822,613 >|245,069 >|336%
+|coachShuttleGathering >|134,522 >|47,455 >|283%
+|conferenceScheduling >|1,566 >|1,534 >|102%
+|curriculumCourse >|56,685 >|54,021 >|105%
+|examination >|28,260 >|20,130 >|140%
+|flightCrewScheduling >|186,906 >|163,184 >|115%
+|investment >|606,111 >|127,119 >|477%
+|machineReassignment >|28,678 >|21,204 >|135%
+|meetingscheduling >|3,034 >|4,181 >|73%
+|nQueens >|419,201 >|309,581 >|135%
+|nurserostering >|33,123 >|17,460 >|190%
+|pas >|59,049 >|79,651 >|74%
+|projectjobscheduling >|128,209 >|41,125 >|312%
+|rockTour >|282,508 >|62,018 >|456%
+|taskAssigning >|31,814 >|15,733 >|202%
+|tennis >|407,459 >|168,607 >|242%
+|travelingtournament >|118,370 >|72,143 >|164%
+|tsp >|435,091 >|284,803 >|153%
+|vehicleRouting >|38,686 >|12,565 >|308%
+|*Average* >|192,008 >|87,777 >|205%
+|===
+
+Bavet is faster than Drools for almost every use case.
+Of course, your mileage may vary.
+Turn on Bavet and if it's not faster in your use case, https://groups.google.com/g/optaplanner-dev[let us know].
+
+Drools and Bavet are both still improving.
+And we have some crazy ideas for a nextgen score calculation implementation
+that uses https://github.com/quarkusio/gizmo[Gizmo]
+to generate domain-specific Bavet-like nodes.
+This performance race is far from over.
+
+== Scaling
+
+Does Bavet scale well?
+
+On commodity hardware, we ran a 5 minutes VRP benchmark on different dataset sizes,
+to compare how Drools and Bavet scale up:
+
+// TODO redo this benchmark on the benchmark machine
+
+// TODO Graph!
+
+.Score calculation speed on different dataset sizes of VRP
+|===
+|Dataset |Drools |Bavet |Speed up
+
+|belgium-n50-k10 >|66,585/s >|264,609/s >|+297.40%
+|belgium-n100-k10 >|50,636/s >|205,620/s >|+306.07%
+|belgium-n500-k20 >|31,431/s >|123,974/s >|+294.43%
+|belgium-n1000-k20 >|20,007/s >|75,556/s >|+277.65%
+|belgium-n2750-k55 >|27,072/s >|81,995/s >|+202.88%
+|*Average* >|39,146/s >|150,350/s >|+284.08%
+|===
+
+Same story, but the performance gap does close as the scale goes up.
+
+== Memory
+
+What about memory? In our non-scientific observations (JFR etc),
```

Comment:
Please provide a link or explain JFR. I had to Google it.

Reply:
Removed the memory section because I haven't run the benchmarks properly to make bold statements around it.